### PR TITLE
Add missing break between polytype quantification and arrow-type body

### LINKED ISSUE #1604: Feature request: improved formatting of wrapped explicit polytypes
The following example formats to itself on current `master`:

```ocaml
let t1 : 'a 'b. 'a t -> b t = ()

let t2 :
      'a 'b.    'a t________________________________
      -> 'b t_______________________________________ =
  ()

let t3 :
      'long 'sequence 'of_ 'universally 'quantified 'type_ 'variables 'that
      'must 'wrap.    'a t_________________________________________________
      -> 'b t______________________________________________________________
      -> 'c t______________________________________________________________ =
  ()

```

The four spaces between the type quantification and the type body in `t2` and `t3` wrong to me: I suspect that it should only be a single space to match the `t1` case. I also wonder if wrapped lines should be indented by some amount (e.g. 2 spaces) to make the scope of the universal variables more clear.

I'm not sure whether the existing behaviour is intentional or is a bug, so have tagged this as a feature request for now. Thanks :-)

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #1604 

Diff with test_branch:
```diff
diff --git a/infer/src/absint/AnalysisCallbacks.ml b/infer/src/absint/AnalysisCallbacks.ml
index 28db09594..840d0d947 100644
--- a/infer/src/absint/AnalysisCallbacks.ml
+++ b/infer/src/absint/AnalysisCallbacks.ml
@@ -11,7 +11,7 @@ module L = Logging
 type callbacks =
   { get_proc_desc_f: Procname.t -> Procdesc.t option
   ; html_debug_new_node_session_f:
-      'a.    ?kind:[`ComputePre | `ExecNode | `ExecNodeNarrowing | `WTO]
+      'a. ?kind:[`ComputePre | `ExecNode | `ExecNodeNarrowing | `WTO]
       -> pp_name:(Format.formatter -> unit) -> Procdesc.Node.t -> f:(unit -> 'a)
       -> 'a
   ; proc_resolve_attributes_f: Procname.t -> ProcAttributes.t option }
diff --git a/infer/src/absint/AnalysisCallbacks.mli b/infer/src/absint/AnalysisCallbacks.mli
index 3a0f569d3..3155ee1f9 100644
--- a/infer/src/absint/AnalysisCallbacks.mli
+++ b/infer/src/absint/AnalysisCallbacks.mli
@@ -31,7 +31,7 @@ val proc_resolve_attributes : Procname.t -> ProcAttributes.t option
 type callbacks =
   { get_proc_desc_f: Procname.t -> Procdesc.t option
   ; html_debug_new_node_session_f:
-      'a.    ?kind:[`ComputePre | `ExecNode | `ExecNodeNarrowing | `WTO]
+      'a. ?kind:[`ComputePre | `ExecNode | `ExecNodeNarrowing | `WTO]
       -> pp_name:(Format.formatter -> unit) -> Procdesc.Node.t -> f:(unit -> 'a)
       -> 'a
   ; proc_resolve_attributes_f: Procname.t -> ProcAttributes.t option }
diff --git a/lib/js_of_ocaml/json.ml b/lib/js_of_ocaml/json.ml
index 727d194e4b..aabc476666 100644
--- a/lib/js_of_ocaml/json.ml
+++ b/lib/js_of_ocaml/json.ml
@@ -25,8 +25,8 @@ class type json =
     method parse : 'a. js_string t -> 'a meth
 
     method parse_ :
-      'a 'b 'c 'd.    js_string t
-      -> ('b t, js_string t -> 'c -> 'd) meth_callback -> 'a meth
+      'a 'b 'c 'd. js_string t -> ('b t, js_string t -> 'c -> 'd) meth_callback
+      -> 'a meth
 
     method stringify : 'a. 'a -> js_string t meth
 
diff --git a/src/dune_engine/string_with_vars.ml b/src/dune_engine/string_with_vars.ml
index e35d9b614..6ed45a8d5 100644
--- a/src/dune_engine/string_with_vars.ml
+++ b/src/dune_engine/string_with_vars.ml
@@ -228,7 +228,7 @@ module Make (A : Applicative_intf.S1) = struct
   module App = Applicative.Make (A)
 
   let partial_expand :
-        'a.    t -> mode:'a Mode.t -> dir:Path.t
+        'a. t -> mode:'a Mode.t -> dir:Path.t
         -> f:Value.t list option A.t expander -> 'a Partial.t A.t =
    fun ({template; syntax_version} as t) ~mode ~dir ~f ->
     match template.parts with
```

### CODE REVIEW

File: lib/Fmt_ast.ml [hunk 1]
```diff
@@ -734,11 +734,12 @@ and fmt_core_type c ?(box = true) ?(in_type_declaration = false) ?pro
...
           fits_breaks ""
             (String.make (Int.max 1 (indent - String.length pro)) ' ')
-      | _ ->
+      | _ when box ->
           fmt_if_k
             Poly.(c.conf.break_separators = `Before)
             (fmt_or_k c.conf.ocp_indent_compat (fits_breaks "" "")
-               (fits_breaks "" "   ") ) )
+               (fits_breaks "" "   ") )
+      | _ -> noop )
```
Comment (Julow) [inline lib/Fmt_ast.ml:742]: This fixes the problem only by chance I think. I get better result by passing `~box:true` in `Ptyp_poly` below, the initial indentation is not lost and I prefer how it breaks.

Comment (gpetiot) [reply lib/Fmt_ast.ml:742]: Yes I've also noticed this, but I think the wrapping was a feature of the original formatting, that's why I preserved it.
Though I agree that breaking makes the output easier to read, I aimed at reducing the diff for the user.

Comment (gpetiot) [reply lib/Fmt_ast.ml]: Didn't receive a reply from Josh so I guess whether or not it was intended we can change it, here is the test_branch diff for the janestreet profile for the record:
```diff
diff --git a/infer/src/absint/AnalysisCallbacks.ml b/infer/src/absint/AnalysisCallbacks.ml
index 95d6fb773..815966a69 100644
--- a/infer/src/absint/AnalysisCallbacks.ml
+++ b/infer/src/absint/AnalysisCallbacks.ml
@@ -11,8 +11,12 @@ module L = Logging
 type callbacks =
   { get_proc_desc_f : Procname.t -> Procdesc.t option
   ; html_debug_new_node_session_f :
-      'a. ?kind:[ `ComputePre | `ExecNode | `ExecNodeNarrowing | `WTO ]
-      -> pp_name:(Format.formatter -> unit) -> Procdesc.Node.t -> f:(unit -> 'a) -> 'a
+      'a.
+      ?kind:[ `ComputePre | `ExecNode | `ExecNodeNarrowing | `WTO ]
+      -> pp_name:(Format.formatter -> unit)
+      -> Procdesc.Node.t
+      -> f:(unit -> 'a)
+      -> 'a
   ; proc_resolve_attributes_f : Procname.t -> ProcAttributes.t option
   }
 
diff --git a/infer/src/absint/AnalysisCallbacks.mli b/infer/src/absint/AnalysisCallbacks.mli
index ad9d31a35..301eb4723 100644
--- a/infer/src/absint/AnalysisCallbacks.mli
+++ b/infer/src/absint/AnalysisCallbacks.mli
@@ -31,8 +31,12 @@ val proc_resolve_attributes : Procname.t -> ProcAttributes.t option
 type callbacks =
   { get_proc_desc_f : Procname.t -> Procdesc.t option
   ; html_debug_new_node_session_f :
-      'a. ?kind:[ `ComputePre | `ExecNode | `ExecNodeNarrowing | `WTO ]
-      -> pp_name:(Format.formatter -> unit) -> Procdesc.Node.t -> f:(unit -> 'a) -> 'a
+      'a.
+      ?kind:[ `ComputePre | `ExecNode | `ExecNodeNarrowing | `WTO ]
+      -> pp_name:(Format.formatter -> unit)
+      -> Procdesc.Node.t
+      -> f:(unit -> 'a)
+      -> 'a
   ; proc_resolve_attributes_f : Procname.t -> ProcAttributes.t option
   }
 
diff --git a/src/dune_engine/string_with_vars.ml b/src/dune_engine/string_with_vars.ml
index 70a828a55..b0b7e67d3 100644
--- a/src/dune_engine/string_with_vars.ml
+++ b/src/dune_engine/string_with_vars.ml
@@ -241,7 +241,11 @@ module Make (A : Applicative_intf.S1) = struct
   module App = Applicative.Make (A)
 
   let partial_expand :
-        'a. t -> mode:'a Mode.t -> dir:Path.t -> f:Value.t list option A.t expander
+        'a.
+        t
+        -> mode:'a Mode.t
+        -> dir:Path.t
+        -> f:Value.t list option A.t expander
         -> 'a Partial.t A.t
     =
    fun ({ template; syntax_version } as t) ~mode ~dir ~f ->
```

File: test/passing/polytypes-janestreet.ml.ref [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+let t2 :
+      'a 'b.
+      'a t________________________________ -> 'b t_______________________________________
+  =
+  ()
+;;
+
+let t3 :
+      'long 'sequence 'of_ 'universally 'quantified 'type_ 'variables 'that 'must 'wrap.
+      'a t_________________________________________________
```
[the remark this answers is not in the document]
Comment (gpetiot) [reply test/passing/polytypes-janestreet.ml.ref:12]: No this is intended to be like this when `ocp-indent-compat=true` (like in the js profile), there is 3 more spaces only when `ocp-indent-compat=false`

**Review by Julow — APPROVED** [review]
Looks good to me then :)

**Comment by gpetiot** [comment]
cc @hhugo to have the opinion of Janestreet users on this change! (don't know who else to ping since @ceastlund is on leave)

**Comment by gpetiot** [comment]
cc @charlesetc @olydis to have the opinion of Janestreet users on this change!

**Comment by charlesetc** [comment]
These changes look good to us, thanks for the improvement!

**Pull request merged**